src/txmempool.cpp

```diff
@@ -57,46 +57,75 @@ size_t CTxMemPoolEntry::GetTxSize() const
-        stageEntries.erase(cit);
-        const setEntries &setChildren = GetMemPoolChildren(cit);
-        for (txiter childEntry : setChildren) {
-            cacheMap::iterator cacheIt = cachedDescendants.find(childEntry);
-            if (cacheIt != cachedDescendants.end()) {
-                // We've already calculated this one, just add the entries for this set
-                // but don't traverse again.
-                for (txiter cacheEntry : cacheIt->second) {
-                    setAllDescendants.insert(cacheEntry);
+    const auto epoch = GetFreshEpoch();
+    // set up the new_cache_line to contain all of our transaction's children (note --
+    // already de-duplicated in case multiple outputs of ours are spent in one
+    // transaction).
+    // Always creates a new cache line entry.
+    vecEntries& new_cache_line = cache[update_it];
+    assert(new_cache_line.empty());
+    // next_idx_to_walk index keeps track of the elements that we've
+    // already expanded.
+    // Invariants:
+    // If index is < next_idx_to_walk, we've already walked it.
+    // If index is >= next_idx_to_walk, we need to walk it.
+    // If next_idx_to_walk >= new_cache_line.size(), we're finished.
+    size_t next_idx_to_walk = 0;
+    txiter next_it = update_it;
+    do {
+        for (const txiter child_it : GetMemPoolChildren(next_it)) {
+            // N.B. children can also be grand children, so guarding
+            // here is neccessary
+            if (visited(child_it)) continue;
+            // if it exists in the cache, copy cached descendants
+            // - All elements in cache are not excluded, no need to check when adding.
+            // - All keys in the cache are already excluded, no need to add
+            //   `child_it` to the new_cache_line (but not all excluded keys
+            //   are in the cache yet!)
+            cacheMap::iterator cached_grand_children = cache.find(child_it);
+            if (cached_grand_children != cache.end()) {
+                for (const txiter grand_child_it : cached_grand_children->second) {
+                    if (visited(grand_child_it)) continue;
+                    // place on the back and then swap into the next_idx_to_walk index
+                    // so we don't walk it ourselves (whoever put the grand
+                    // child in the cache must have already traversed this, so it must
+                    // not be excluded)
+                    new_cache_line.emplace_back(grand_child_it);
+                    std::swap(new_cache_line[next_idx_to_walk], new_cache_line.back());
+                    ++next_idx_to_walk;
                 }
-            } else if (!setAllDescendants.count(childEntry)) {
+            } else {
                 // Schedule for later processing
-                stageEntries.insert(childEntry);
+                // Element must not be excluded, or it would be in the cache already.
+                new_cache_line.emplace_back(child_it);
             }
```

Comment:
The internal loop seems redundant. Could this work:
```
    do {
        for (const txiter child_it : GetMemPoolChildren(next_it)) {
            if (visited(child_it)) continue;
            new_cache_line.emplace_back(child_it);
        }

        // finish loop here!
        if (next_idx_to_walk >= new_cache_line.size()) break;

        // Prepare for next iteration:
        next_it = new_cache_line[next_idx_to_walk];
        ++next_idx_to_walk;
    } while (true);
```
?

Reply:
No because then we aren't using the cached traversal at all.

It could be correct, but has some algorithmic complexity blowups.

Reply:
You update is still not using the cache lines....

Reply:
> It could be correct, but has some algorithmic complexity blowups.

How it could be, if both every element is visited once only?

Reply:
I'm saying "could be" because I haven't bothered to actually check your code above for correctness, given the glaring issue that it doesn't use the caches which leads to unacceptable algorithmic complexity.

Reply:
I guess to pull up -- I do agree that the loop is a bit atypical in structure, but I do think that it's the most straightforward to reason about because it's pretty critical when certain checks or variable updates happen.

Reply:
> "... it doesn't use the caches..."

Corrected:
```
    txiter next_it = update_it;
    for (size_t next_idx_to_walk = 0; next_idx_to_walk < new_cache_line.size(); ++next_idx_to_walk) {
        for (const txiter child_it : GetMemPoolChildren(next_it)) {
            if (visited(child_it)) continue;
            if (cache.find(child_it) != cache.end()) continue;
            new_cache_line.emplace_back(child_it);
        }
        next_it = new_cache_line[next_idx_to_walk];
    }
```

Reply:
So now the code checks if there is a cache, and then continues... but you're now both not using the cache, and incorrect. If you find a cache line for a child then you must include all of it's cache entries into the new_cache_line otherwise a future call will skip updating descendants properly.

The code that is in the PR is (I believe) correct & does not have superfluous lines. If there was a good simplification of that logic I'd be for it, but I think that it already sort of is at the minimal correct form. There's not an extra internal loop to get rid of.

Reply:
Code in the PR passes the test from #18485. 
All my suggestions fail the test.